File: registry/tools/calculator/calculator.ts
/**
 * Calculator tools for basic arithmetic operations.
 */

import { llm } from "mirascope";

/**
 * Add two numbers together.
 */
export const add = llm.tool({
  name: "add",
  description: "Add two numbers together.",
  parameters: {
    type: "object",
    properties: {
      a: { type: "number", description: "The first number." },
      b: { type: "number", description: "The second number." },
    },
    required: ["a", "b"],
  },
  execute: ({ a, b }: { a: number; b: number }) => a + b,
});

/**
 * Subtract b from a.
 */
export const subtract = llm.tool({
  name: "subtract",
  description: "Subtract b from a.",
  parameters: {
    type: "object",
    properties: {
      a: { type: "number", description: "The number to subtract from." },
      b: { type: "number", description: "The number to subtract." },
    },
    required: ["a", "b"],
  },
  execute: ({ a, b }: { a: number; b: number }) => a - b,
});

/**
 * Multiply two numbers.
 */
export const multiply = llm.tool({
  name: "multiply",
  description: "Multiply two numbers.",
  parameters: {
    type: "object",
    properties: {
      a: { type: "number", description: "The first number." },
      b: { type: "number", description: "The second number." },
    },
    required: ["a", "b"],
  },
  execute: ({ a, b }: { a: number; b: number }) => a * b,
});

/**
 * Divide a by b.
 */
export const divide = llm.tool({
  name: "divide",
  description: "Divide a by b.",
  parameters: {
    type: "object",
    properties: {
      a: { type: "number", description: "The dividend." },
      b: { type: "number", description: "The divisor." },
    },
    required: ["a", "b"],
  },
  execute: ({ a, b }: { a: number; b: number }) => {
    if (b === 0) {
      throw new Error("Cannot divide by zero");
    }
    return a / b;
  },
});

File: registry/tools/calculator/calculator.py
"""Calculator tools for basic arithmetic operations."""

from mirascope import llm


@llm.tool
def add(a: float, b: float) -> float:
    """Add two numbers together.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The sum of a and b.
    """
    return a + b


@llm.tool
def subtract(a: float, b: float) -> float:
    """Subtract b from a.

    Args:
        a: The number to subtract from.
        b: The number to subtract.

    Returns:
        The difference of a and b.
    """
    return a - b


@llm.tool
def multiply(a: float, b: float) -> float:
    """Multiply two numbers.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The product of a and b.
    """
    return a * b


@llm.tool
def divide(a: float, b: float) -> float:
    """Divide a by b.

    Args:
        a: The dividend.
        b: The divisor.

    Returns:
        The quotient of a divided by b.

    Raises:
        ValueError: If b is zero.
    """
    if b == 0:
        raise ValueError("Cannot divide by zero")
    return a / b
